### PR TITLE
Generic fri folder

### PR DESCRIPTION
**Pull request by rjeli**
In preparation for circle-fri. Based off my matrix-rework branch, so drafting until that's merged, but it's good to go otherwise.

### CODE REVIEW

**Comment by rjeli** [comment]
Hm, seems like I can't change the base branch after creating the PR. So you can take a look but I'll create another PR into main after `matrix-rework` is merged

**Comment by dlubarov** [comment]
I think it may change automatically once that lands, but new PR is fine too if not

File: fri/src/config.rs [hunk 1]
```diff
@@ -11,3 +18,28 @@ impl<M> FriConfig<M> {
...
+        m.par_rows()
+            .enumerate()
+            .map(|(i, r)| Self::fold_row(i, log_height, beta, r))
+            .collect()
+    }
+
+    /// Mix in evaluations from an input at this log_height.
+    fn mix_in(&self, index: usize, log_height: usize, current: &mut F, new: F);
+    fn mix_in_column(&self, current: &mut [F], new: &[F]) {
+        let log_height = log2_strict_usize(current.len());
```
Comment (dlubarov) [inline fri/src/config.rs:38]: nit: let's `debug_assert!` that the other height matches?

File: fri/src/prover.rs [hunk 2]
```diff
@@ -103,17 +105,20 @@ where
...
     for log_folded_height in (config.log_blowup..log_max_height).rev() {
-        let leaves = RowMajorMatrix::new(current.clone(), 2);
+        let leaves = RowMajorMatrix::new(current, 2);
         let (commit, prover_data) = config.mmcs.commit_matrix(leaves);
         challenger.observe(commit.clone());
-        commits.push(commit);
-        data.push(prover_data);
 
         let beta: F = challenger.sample();
-        current = fold_even_odd(current, beta);
```
Comment (dlubarov) [inline fri/src/prover.rs:113]: Can `fold_even_odd` be removed now?

**Review by dlubarov — APPROVED** [review]
LGTM

**Pull request closed without merging**